src/cio_websocket.c

```diff
@@ -156,47 +162,51 @@ static void close_timeout_handler(struct cio_timer *timer, void *handler_context
 
 static void send_close_frame(struct cio_websocket *ws, enum cio_websocket_status_code status_code, struct cio_write_buffer *reason)
 {
+	ws->close_status = cio_htobe16(status_code);
+
 	struct cio_write_buffer wbh;
 	cio_write_buffer_head_init(&wbh);
 	cio_write_buffer_queue_tail(&wbh, &ws->wb_close_status);
 	if (reason != NULL) {
 		cio_write_buffer_splice(reason, &wbh);
 	}
 
-	ws->close_status = cio_htobe16(status_code);
-	enum cio_error err = cio_timer_init(&ws->close_timer, ws->loop, NULL);
-	if (likely(err == CIO_SUCCESS)) {
-		err = ws->close_timer.expires_from_now(&ws->close_timer, close_timeout_ns, close_timeout_handler, ws);
+	if (ws->ws_flags.self_initiated_close == 1) {
+		enum cio_error err = cio_timer_init(&ws->close_timer, ws->loop, NULL);
 		if (likely(err == CIO_SUCCESS)) {
-			send_frame(ws, &wbh, CIO_WEBSOCKET_CLOSE_FRAME, true, close_frame_written);
+			err = ws->close_timer.expires_from_now(&ws->close_timer, close_timeout_ns, close_timeout_handler, ws);
+			if (likely(err == CIO_SUCCESS)) {
+				send_frame(ws, &wbh, CIO_WEBSOCKET_CLOSE_FRAME, true, close_frame_written);
+				return;
+			}
 		}
-	} else {
-		send_frame(ws, &wbh, CIO_WEBSOCKET_CLOSE_FRAME, true, do_nothing);
-		close(ws);
 	}
+
+	send_frame(ws, &wbh, CIO_WEBSOCKET_CLOSE_FRAME, true, do_nothing);
+	close(ws);
 }
 
 static void handle_binary_frame(struct cio_websocket *ws, uint8_t *data, uint64_t length, bool last_frame)
 {
-	if (likely(ws->onbinaryframe_received != NULL)) {
-		ws->onbinaryframe_received(ws, data, length, last_frame);
+	if (likely(ws->onbinaryframe != NULL)) {
+		ws->onbinaryframe(ws, data, length, last_frame);
 	} else {
 		// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
 	}
 }
 
 static void handle_text_frame(struct cio_websocket *ws, uint8_t *data, uint64_t length, bool last_frame)
 {
-	if (likely(ws->ontextframe_received != NULL)) {
-		ws->ontextframe_received(ws, (char *)data, length, last_frame);
+	if (likely(ws->ontextframe != NULL)) {
+		ws->ontextframe(ws, (char *)data, length, last_frame);
 	} else {
 		// TODO: handle_error(s, WS_CLOSE_UNSUPPORTED);
 	}
 }
 
 static int handle_close_frame(struct cio_websocket *ws, uint8_t *data, uint64_t length)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to not always return the same value. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS3516)